File: unity/Assets/Scripts/Multiplayer/Games/GameDataResponse.cs
using Michitai.Multiplayer.Errors;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace Michitai.Multiplayer.Games
{
    /// <summary>
    /// Response containing game data with typed deserialization support in Unity.
    /// Uses Unity's JsonUtility for deserialization.
    /// </summary>
    /// <typeparam name="T">The type to deserialize the game data into.</typeparam>
    [System.Serializable]
    public class GameDataResponse<T> : ApiResponse<EGameDataGameGetError> where T : class, new()
    {
        [SerializeField]
        private string data_json;

        /// <summary>
        /// The type identifier for the game data.
        /// </summary>
        public string type;

        /// <summary>
        /// The ID of the game.
        /// </summary>
        public int game_id;

        /// <summary>
        /// The deserialized game data object (parsed from JSON string).
        /// </summary>
        public T Data
        {
            get
            {
                return JsonUtility.FromJson<T>(data_json);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/games/GameDataResponse.java
package com.michitai.multiplayer.games;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.michitai.multiplayer.ApiResponse;

/**
 * Response containing game data with typed deserialization support.
 * Uses Jackson for deserialization.
 *
 * @param <T> The type to deserialize the game data into.
 */
public class GameDataResponse<T> extends ApiResponse {
    @JsonProperty("type")
    private String type;

    @JsonProperty("game_id")
    private int gameId;

    @JsonProperty("data_json")
    private String dataJson;

    @JsonProperty("data")
    private T data;

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public int getGameId() {
        return gameId;
    }

    public void setGameId(int gameId) {
        this.gameId = gameId;
    }

    public String getDataJson() {
        return dataJson;
    }

    public void setDataJson(String dataJson) {
        this.dataJson = dataJson;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }
}
